### ISSUE TITLE
Yaml Improvements for Version 0.6

### ISSUE THREAD

**Issue by bradrydzewski** [drone/drone#1505]
__THIS IS NOT__ an open invitation for to propose a completely different yaml format. The purpose of this issue is to solve some specific deficiencies in the yaml format. __Do not submit a proposal if it does not meet the below goals__.

__Goal 1, Mix plugins and shell commands__

The purpose of this exercise is to give people the ability to mix plugins with their build steps. Right now plugins are only executed after all build steps complete, however, it is possible someone might want to execute a plugin, and then execute more shell commands after that.

__Goal 2, Execute plugins on failure__

Right now plugins are not executed if the build fails. In some cases, such as reporting and artifacts, it is necessary to execute commands after the build completes even on failure.

__Goal 3, Security__

If we are going to execute plugins as part of the build we absolutely need to think about security. There is a subset of plugins (pypi, git push, etc) that could provide an attack vector for a malicious pull request to expose tokens and ssh private keys. This needs to be accounted for:
https://github.com/drone/drone/issues/1476#issuecomment-182635837

__Goal 4, Remove Publish, Deploy sections__

The notion of publish and deploy sections are too rigid and confusing and need to be removed. We frequently get the question "how can I deploy with a shell command". This of course can be done directly in the build section, however, the fact that we have a deploy section confuses people into thinking deployments can only happen in the deploy section (understandably).

__Goal 5, Step syntax cannot change__

The below structure of an individual step __cannot be changed__. This is non-negotiable

```Go
type Step struct {
	Image          string
	Pull           bool
	Privileged     bool
	Environment    MapEqualSlice
	Entrypoint     Command
	Command        Command
	ExtraHosts     []string `yaml:"extra_hosts"`
	Volumes        []string
	Devices        []string
	Net            string
	AuthConfig     AuthConfig `yaml:"auth_config"`
	Memory         int64      `yaml:"mem_limit"`
	CPUSetCPUs     string     `yaml:"cpuset"`
	OomKillDisable bool       `yaml:"oom_kill_disable"`
}
```

__Goal 6, backward compatibility__

We will keep the existing yaml format for backward compatibility. We will likely introduce a version attribute in the yaml to specify the version:

```
version: 2
```

__BONUS move secrets to a plugin__

Right now we inject secrets into the yaml with a `$$` notation. Another approach might be to mount secrets to the filesystem similar to kubernetes. It was suggested that perhaps we could delegate this to plugins, which I think would be very interesting, but might merit a completely separate issue. Food for thought ....

**Comment by bradrydzewski** [comment]
This is an example yaml implementation that is similar to what we have today. This was an example that I've been considering:

```yaml
pipeline:
  clone:
    plugin: git

  cache:
    plugin: cache
    mount: [ node_modules ]

  mysql:
    image: mysql
    detach: true

  postgres:
    image: mysql
    detach: true

  unit_tests:
    image: golang
    commands:
      - go build
      - go get
      - go test

  prod_deploy:
    plugin: heroku
    app: foo
    token: $$HEROKU_TOKEN
    when:
      event: [ push, tag ]
      branch: master

catch:
  rollback:
    plugin: heroku_rollback
    token: $$HEROKU_TOKEN
    app: foo

finally:
  notify:
    plugin: slack
    token: $$SLACK_TOKEN
```

__Things I Like__

1. maintains the same step syntax
2. removes the publish, deploy, notify sections and executes everything in the pipeline section
3. pipeline section lets you declare plugins and shell steps
4. catch / finally section of the yaml lets you run steps on failure. catch / finally is something developers are used to and should minimize learning curve

__Questions I Still Have__

1. docker compose has a separate `services` section. Should we mirror this? Or launch service containers from the pipeline section with `detach: true` as seen above?
2. we might still need to enforce some level of ordering for a subset of plugins. The clone and cache plugins need to run at the start of the build, before subsequent steps
3. mixing plugins and commands makes security for pull requests a challenge. We would likely not want to execute any plugins if the yaml changed (checksum fails). Some plugins (git clone, cache) are still necessary. Can we handle this without having to manage a white list? Other ideas?

**Comment by gtaylor** [comment]
> docker compose has a separate services section. Should we mirror this? Or launch service containers from the pipeline section with detach: true as seen above?

I think I like a separate services section. I think this mostly comes down to a case of cosmetics, though.

> we might still need to enforce some level of ordering for a subset of plugins. The clone and cache plugins need to run at the start of the build, before subsequent steps

That makes sense. We can write some helpful error messages or link to a specific section of the docs for more on fixing such a broken pipeline.

> mixing plugins and commands makes security for pull requests a challenge. We would likely not want to execute any plugins if the yaml changed (checksum fails). Some plugins (git clone, cache) are still necessary. Can we handle this without having to manage a white list? Other ideas?

As a first draft, I'd give people the shotgun, tell them how to use it safely, and see what happens. A whitelist is possible and maybe even necessary, but it does introduce some unexpected behavior for those who are new. It's an inconsistency or an exception to the rule that people will need to learn.

By using `when` tags (and encouraging such a thing), people can opt into such protections. If `when` isn't enough, we might be able to add other filters.

**Comment by bradrydzewski** [comment]
> I think I subjectively like a separate services section. This mostly comes down to a case of cosmetics, though. Real scientific, I know.

The `services` section is new in the docker-compose v2 yaml file which didn't exist a few weeks ago when I had posted the above sample in a different thread. I do think having commonality with the docker-compose yaml has some benefits, including the possibility to use the same file for drone + docker compose in the future.

> As a first draft, I'd give people the shotgun, tell them how to use it safely, and see what happens. A whitelist is possible and maybe even necessary

I think to do this we need to change a subset of plugins that don't require authentication, such as the ssh, rsync and git push plugins. These plugins will use the default ssh key associated with your drone account, which means someone could add an ssh step without secrets, and without the checksum protection.

I think using the ssh key by default was convenient, however, it is something we should eliminate in the plugins. These plugins should require an ssh key is provided in the `.drone.sec` file like all other plugins.

This would solve the issue entirely because it would force the use of a .drone.sec file for all plugins (except clone and cache) and therefore come with the protections of the checksum.

**Comment by tboerger** [comment]
I think a separate service section makes sense. As we have to detach it anyway it's easier to understand because of compose.

Maybe we should keep separate sections for cache and clone like before? That way nobody can break the pipeline and we can enforce the order for them even easier

**Comment by bradrydzewski** [comment]
> As a first draft, I'd give people the shotgun, tell them how to use it safely, and see what happens.

@gtaylor agreed, however, there are still some security issues we'll need to resolve. The proposed yaml design would run plugins during pull requests and would be subject to the vectors described in #1476. The proposed workaround for #1475 was to only execute plugins when the checksum is verified. This can sort-of work in the above yaml structure but we'll need to make sure (for example) we don't skip the clone plugin when the checksum fails.

**Comment by skolodyazhnyy** [comment]
@bradrydzewski Nice proposal, looking forward to try it. It seems to be a little bit more verbose and than current format but definitely way more flexible. Just wondering, are you considering possibility to run some steps in parallel?

**Comment by bradrydzewski** [comment]
I've been thinking a bit more about some of the security concerns (see https://github.com/drone/drone/issues/1476 and https://github.com/drone/drone/issues/1145) and how they can be addressed to enable safe plugin execution during the build and for pull requests.

I propose that when we create the `.drone.sec` file we specify which secrets receive which plugins:

```
checksum: 688787d8ff144c502c7f

plugins:
  docker:
    email: foo@bar.com
    username: foo
    password: baz
```

This also would impact our ability to use string interpolation, since we need to control when and where secrets are injected. Instead of interpolation we could mount secrets to the filesystem similar to kubernetes [1][2] which should simplify our yaml files.

There are currently three different types of secrets:

* **source secrets** encrypted in the .drone.sec file
* **repo secrets** provided to each build to clone from the remote system (ie ssh key)
* **global secrets** set by the administrator and injected at a global level for all builds

The source secrets will be defined in the .drone.sec file, as mentioned above, per plugin. The repo secrets will also be defined in a similar fashion, but instead by drone itself, per plugin. This means drone will come up with a list of which plugins can, by default, receive ssh keys or netrc files. There will be very very few plugins that drone automatically sends secrets to, which means aside from the git and hg plugin you should almost always use the .drone.sec file.
 
If we are explicit about which plugins are provided which secrets it makes it much easier to prevent secrets from being exposed to malicious plugins. Also by avoiding interpolation we can prevent people from accidentally echoing secrets to the console.

More details about exact format of `.drone.sec` file once I've had more time to read more about kubernetes and (hopefully) align with their approach.

[1] http://kubernetes.io/v1.1/docs/user-guide/secrets.html
[2] http://kubernetes.io/v1.1/docs/user-guide/images.html#specifying-imagepullsecrets-on-a-pod

**Comment by bradrydzewski** [comment]
closing since this has been available in upcoming 0.5 and seems to work quite well based on (limited) feedback. We can tweak as needed in subsequent releases.
